### PR TITLE
Assignment1

### PR DESCRIPTION
**Pull request by MattJett-EWU**
- Added global suppression file in ../SecretSanta.Api/
- Used Directory.Build.props file to enforce global TargetFramework and Nullable attributes to whole solution.

### CODE REVIEW

**Comment by MattJett-EWU** [comment]
holy crap I hope I did this right...

**Comment by kveselits** [comment]
I will review this.

**Comment by kveselits** [comment]
Turn on nullability in the SecretSanta.API project ✔

Configure nullability in SecretSanta.API project to be conditional based Nullability not already being set (pending Thursday lecture). ✔

> I think this was supposed to be configured specifically for the SecretSanta.Api project only. I could be wrong on this. Feel free to correct me if I'm wrong (I wasn't 100% sure about this part myself).

Add the following classes to the SecretSanta.Business project with the following the associated properties:

Gift ✔

int Id as read-only ✔
string Title ✔
string Description ✔
string Url ✔
User User ✔

User ✔

int Id as read-only ✔
string FirstName ✔
string LastName ✔
<a collection of> Gifts ✔
Add non-default constructors for Gift and User. ✔

Add unit tests to all projects except SecretSanta.Web and fully unit test the new classes in SecretSanta.Business(pending Thursday lecture). ✔

Refactor nullability setting into solution level props and targets files (pending Thursday lecture). ✔

> Like I mentioned before, I'm not sure if the conditional nullability was meant to be made solution level. I assumed the instructions meant only the `<nullable>enable</nullable>` property. Like I said I might be wrong and hopefully someone will correct me if I'm mistaken.

Add the following list of code analysis assemblies and appropriately handle all warnings: IntelliTectAnalyzer.dll,Microsoft.NetCore.Analyzers.dll,Microsoft.CodeQuality.Analyzers.dll, and Microsoft.NetCore.Analyzers.dll. (Adding the IntelliTect one will automatically add the others since they are dependencies on the IntelliTect one.) ✔

Refactor out a solution level global suppression file for disabling code analysis warnings across all projects (to do this add a reference to the same suppression file(s) from all projects). ✔

> Just a quick note that a couple of your suppressions have the justification still `<Pending>`. Also, you probably wouldn't have been able to implement this since it was covered in the lecture after the assignment was due, but if you remove the `Scope` and `Target` parameters you won't need to have duplicate suppressions for the same warning.

Configure Azure DevOps build for continuous integration to compile and run all unit tests ✔

If any updates occur in Assignment1 prior to your PR, rebase onto Assignment1. Finally, and just prior to submitting your PR, rebase from master (which has additional commits added after Assessment1 was created such as the addition of the cSpell.json in commit #7b106a6). ✔

Extra Credit
The following are options for extra credit (you don't need to do them all):

Use reflection to test all properties on all classes in SecretSanta.Business. ❌✔
Configure tests to run in parallel both locally and in Azure DevOps and be sure the both the test and SUT code is appropriately thread safe for the tests to pass reliably. ❌✔

> I honestly don't know how to check if you implemented the last one so I just left this section unchanged.

> Overall I think your code looks awesome and is thoughtfully and consistently organized between different classes. The code is definitely intent revealing and I didn't feel like I needed a comment to explain any of it. Your null validation on all of your properties looks great and I feel like you really went the extra mile here.

**Comment by MattJett-EWU** [comment]
> Just a quick note that a couple of your suppressions have the justification still `<Pending>`. Also, you probably wouldn't have been able to implement this since it was covered in the lecture after the assignment was due, but if you remove the `Scope` and `Target` parameters you won't need to have duplicate suppressions for the same warning.

>> Can you help me understand the `<Pending>` attribute value? VS auto placed these in like this and it seemed to be working as intended, so I left the `<Pending>` part alone. And I did remove the `Scope` and `Target` attributes. Only the first 4ish lines contained a `Scope` and `Target` by design. I wanted the underscore suppression to only apply to the unit tests as well as the string uri suggestion suppression for the Url Gift property that only needs to apply to that class. But, everything else there you can see does not contain a `Scope` and `Target`. If you see differently, then please let me know, because I'm worried now that it is not reflecting to others as I see it in my repo.

> Overall I think your code looks awesome and is thoughtfully and consistently organized between different classes. The code is definitely intent revealing and I didn't feel like I needed a comment to explain any of it. Your null validation on all of your properties looks great and I feel like you really went the extra mile here.

>> Thank you, appreciate the positive feedback here.

**Comment by kveselits** [comment]
> > > Can you help me understand the `<Pending>` attribute value? VS auto placed these in like this and it seemed to be working as intended, so I left the `<Pending>` part alone. And I did remove the `Scope` and `Target` attributes. Only the first 4ish lines contained a `Scope` and `Target` by design. I wanted the underscore suppression to only apply to the unit tests as well as the string uri suggestion suppression for the Url Gift property that only needs to apply to that class. But, everything else there you can see does not contain a `Scope` and `Target`. If you see differently, then please let me know, because I'm worried now that it is not reflecting to others as I see it in my repo.

Hey Matt, sorry for the late reply I didn't notice your comment on here. 

Don't worry, the suppressions without the Scope and Target are reflecting on my end, I just missed that somehow. It makes sense that you were intending for those warnings to target those specific classes. You can disregard that part of my suggestion as I wasn't sure if you were aware of the ability to target the entire solution.

Mark covered the Justification = "<Pending>" briefly in one of his lectures. It is simply a justification for why you are suppressing the warning. So for the one of the justifications, you could simply write something to the effect of "I suppressed this warning because of [enter reason here]"

File: SecretSanta/src/SecretSanta.Business/Gift.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+			User = user ?? throw new ArgumentNullException(nameof(user));
+		}
+		public Gift() : this(0, "<default_non-null_title>", "", "", new User())
+		{}
+		#endregion
+
+
+		#region Private Helper Methods
+		// Reference: https://github.com/IntelliTect-Samples/EWU-CSCD379-2020-Winter/blob/Lecture-2020.01.09/SecretSanta/src/SecretSanta.Business/BlogPost.cs
+		private string AssertIsNotNullOrWhitespace(string value)
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Business/Gift.cs]: Member can be made static

File: SecretSanta/src/SecretSanta.Business/User.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			Gifts = gifts ?? throw new ArgumentNullException(nameof(gifts));
+		}
+		public User() : this(0, "<default_non-null_firstname>", "<default_non-null_lastname>", new List<Gift>())
+		{}
+		#endregion
+
+
+		#region Private Helper Methods
+		// Reference: https://github.com/IntelliTect-Samples/EWU-CSCD379-2020-Winter/blob/Lecture-2020.01.09/SecretSanta/src/SecretSanta.Business/BlogPost.cs
+		private string AssertIsNotNullOrWhitespace(string value)
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Business/User.cs]: Member can be made static

File: SecretSanta/_test/SecretSanta.Business.Tests/GiftTests.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            Assert.AreEqual(id, gift.Id, "ID was unexpected");
+            Assert.AreEqual(title, gift.Title, "Title value was unexpected");
+            Assert.AreEqual(description, gift.Description, "Description content was unexpected");
+            Assert.AreEqual(url, gift.Url, "Url value was unexpected");
+            Assert.AreEqual(user, gift.User, "User not found");
+        }
+
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        [DataRow("", "test description", "www.url.com")]
```
Comment (Keboo) [inline SecretSanta/_test/SecretSanta.Business.Tests/GiftTests.cs]: This is good data driven test, but if you are only going to have a single `DataRow` it makes more sense to make it a regular unit test.

File: SecretSanta/src/SecretSanta.Business/BlogPost.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Business/BlogPost.cs]: Did you mean to include this file as well?

File: SecretSanta/src/SecretSanta.Business/SecretSanta.Business.csproj [hunk 5]
```diff
@@ -1,11 +1,24 @@
...
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+	<TargetFramework>netcoreapp3.1</TargetFramework>
   </PropertyGroup>
 
   <ItemGroup>
-    <ProjectReference Include="..\SecretSanta.Data\SecretSanta.Data.csproj" />
+	<Compile Remove="BlogPost.cs" />
```
Comment (Keboo) [inline SecretSanta/src/SecretSanta.Business/SecretSanta.Business.csproj]: Looks like you tried to remove it from VS, but this doesn't actually remove the file from disk, only from the project file. Same with the line below.

File: azure-pipelines.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    restoreSolution: '$(solution)'
+
+- task: VSBuild@1
+  inputs:
+    solution: '$(solution)'
+    msbuildArgs: '/p:DeployOnBuild=true /p:WebPublishMethod=Package /p:PackageAsSingleFile=true /p:SkipInvalidConfigurations=true /p:PackageLocation="$(build.artifactStagingDirectory)"'
+    platform: '$(buildPlatform)'
+    configuration: '$(buildConfiguration)'
+
+- task: VSTest@2
```
Comment (Keboo) [inline azure-pipelines.yml]: Because you have the `DotNetCoreCLI@2` task to run unit tests below you don't need this one.

**Pull request closed without merging**